crates/task-impls/src/transactions.rs

```diff
@@ -233,18 +235,44 @@ where
 
                 drop(consensus);
 
-                let mut block = <TYPES as NodeType>::StateType::next_block(None);
                 let txns = self.wait_for_transactions(parent_leaf).await?;
 
-                for txn in txns {
-                    if let Ok(new_block) = block.add_transaction_raw(&txn) {
-                        block = new_block;
-                        continue;
+                debug!("Prepare VID shares");
+                if !txns.is_empty() {
+                    // TODO https://github.com/EspressoSystems/HotShot/issues/1686
+                    let srs = hotshot_types::data::test_srs(NUM_STORAGE_NODES);
+                    let vid = VidScheme::new(NUM_CHUNKS, NUM_STORAGE_NODES, &srs).unwrap();
+                    // TODO https://github.com/EspressoSystems/jellyfish/issues/375
+                    let mut txns_flatten = Vec::new();
+                    for txn in &txns {
+                        txns_flatten.extend(txn.bytes());
                     }
+                    let vid_disperse = vid.disperse(&txns_flatten).unwrap();
+                    let block = VIDBlockPayload::new(txns, vid_disperse.commit);
+
+                    self.event_stream
+                        .publish(SequencingHotShotEvent::BlockReady(block.clone(), view + 1))
+                        .await;
+
+                    self.event_stream
+                        .publish(SequencingHotShotEvent::VidDisperseSend(
```

Comment:
Should we publish `VidDisperseSend` here, or later in the DA task? Note that `DAProposalSend` is still published in the DA task, and `VidDisperseSend` is intended to be a mirror of `DAProposalSend` for VID. ie. the leader should send the DA proposal at the same time he sends VID disperse data.

If we move it to DA task then the we'll need to add the VID shares to the `BlockReady` event. That raises another question: should we limit the volume of data passed through events? eg. do we do something silly in event processing that takes O(n) time like serialize?

We could address these questions in a future issue, in which case a simple TODO would suffice here.

Reply:
Any opinion here @bfish713 ? I'm not yet familiar with the philosophy behind the event model and the purpose of `BlockReady` .